app/components/graph-editor/canvas.ts

```diff
@@ -379,95 +410,153 @@ export function getStraightEdgePoints(
-        pts.push(e.source.y + shiftPt[1]);
+        pts.push([
+            e.source.x + shiftPt[0],
+            e.source.y + shiftPt[1]
+        ]);
         u[0] *= -1;
         u[1] *= -1;
         shiftPt = getEdgePtShift(u, e.destination, dstDim);
-        pts.push(e.source.x + v[0] + shiftPt[0]);
-        pts.push(e.source.y + v[1] + shiftPt[1]);
+        pts.push([
+            e.source.x + v[0] + shiftPt[0],
+            e.source.y + v[1] + shiftPt[1]
+        ]);
     }
     else if (e.source && !e.destination) {
         console.assert(pt, "error getStraightEdgePoints: pt undefined");
         console.assert(srcDim, "error getStraightEdgePoints: srcDim undefined");
+        let p = pt as number[];
         let v = [
-            (pt as number[])[0] - e.source.x,
-            (pt as number[])[1] - e.source.y
+            p[0] - e.source.x,
+            p[1] - e.source.y
         ];
         let d = MathEx.mag(v);
         let u = [v[0] / d, v[1] / d];
         let shiftPt = getEdgePtShift(u, e.source, srcDim);
-        pts.push(e.source.x + shiftPt[0]);
-        pts.push(e.source.y + shiftPt[1]);
-        pts.push((pt as number[])[0]);
-        pts.push((pt as number[])[1]);
+        pts.push([
+            e.source.x + shiftPt[0],
+            e.source.y + shiftPt[1]
+        ]);
+        pts.push(p);
     }
     else if (!e.source && e.destination) {
         console.assert(pt, "error getStraightEdgePoints: pt undefined");
         console.assert(dstDim, "error getStraightEdgePoints: dstDim undefined");
+        let p = pt as number[];
         let v = [
-            e.destination.x - (pt as number[])[0],
-            e.destination.y - (pt as number[])[1]
+            e.destination.x - p[0],
+            e.destination.y - p[1]
         ];
         let d = MathEx.mag(v);
         let u = [-v[0] / d, -v[1] / d];
-        pts.push((pt as number[])[0]);
-        pts.push((pt as number[])[1]);
+        pts.push(p);
         let shiftPt = getEdgePtShift(u, e.destination, dstDim);
-        pts.push((pt as number[])[0] + v[0] + shiftPt[0]);
-        pts.push((pt as number[])[1] + v[1] + shiftPt[1]);
+        pts.push([
+            p[0] + v[0] + shiftPt[0],
+            p[1] + v[1] + shiftPt[1]
+        ]);
     }
+    pts.push([
+        (pts[0][0] + pts[1][0]) / 2,
+        (pts[0][1] + pts[1][1]) / 2
+    ]);
     return pts;
 }
 
 /**
  * getLoopEdgePoints  
- *   Gets the edge points of a self-referencing node.
+ *   Gets the edge points and midpoint of a self-referencing node.
  */
-export function getLoopEdgePoints(e: DrawableEdge): number[] {
-    // TODO:
-    return [];
+export function getLoopEdgePoints(
+    e: DrawableEdge,
+    src: DrawableNode,
+    srcDim: any
+): number[][] {
+    let u = [MathEx.SIN_22_5, -MathEx.COS_22_5];
+    let v = [-MathEx.SIN_22_5, -MathEx.COS_22_5];
+    let pt0 = getEdgePtShift(u, src, srcDim);
+    let pt1 = getEdgePtShift(v, src, srcDim);
+    let pt2 = [
```

Comment:
The code you are referring to is only used in `getLoopEdgePoints`. Putting it into a function would probably save three lines, but expand the file by more to take into account the header comment for the function. It wouldn't make sense to use the new function outside of `getLoopEdgePoints`, because it is specific to the control points `pt2` and `pt3`. I could make the new function more general, but that would require adding more general functions in order to do what is essentially a vector scale and translation. I don't think expanding the stack for such a calculation would be worth it. If your main concern is readability, note that there is very minimal inline documentation. The full implementation of graph editor is not complete, and I intend to fully document once all of the other minor issues are addressed.